src/sql/reuse/relaxed.rs

```diff
@@ -0,0 +1,147 @@
+///
+/// Finkelstein reuse would be conditional on the order the queries are added,
+/// since 1) is a direct extension of 2), but not the other way around.
+///
+/// This relaxed version of the reuse algorithm considers cases where a query might
+/// reuse just a prefix of another. First, it checks if the queries perform the
+/// same joins, then it checks predicate implication and at last, checks group
+/// by compatibility.
+/// If all checks pass, them the algorithm works like Finkelstein and the query
+/// is a direct extension of the other. However, if not all, but at least one
+/// check passes, then the algorithm returns that the queries have a prefix in
+/// common.
+pub struct Relaxed;
+
+impl ReuseConfiguration for Relaxed {
+    fn reuse_candidates<'a>(qg: &QueryGraph, query_graphs: &'a HashMap<u64, (QueryGraph, MirQuery)>) -> Vec<(ReuseType, &'a QueryGraph)>{
+        let mut reuse_candidates = Vec::new();
+        for &(ref existing_qg, _) in query_graphs.values() {
+            if existing_qg
+                .signature()
+                .is_weak_generalization_of(&qg.signature())
+            {
+                match Self::check_compatibility(&qg, existing_qg) {
+                    Some(reuse) => {
+                        // QGs are compatible, we can reuse `existing_qg` as part of `qg`!
+                        reuse_candidates.push((reuse, existing_qg));
+                    }
+                    None => (),
+                }
+            }
+        }
+
+        reuse_candidates
+    }
+}
+
+impl Relaxed {
+    fn check_compatibility(new_qg: &QueryGraph, existing_qg: &QueryGraph) -> Option<ReuseType> {
+        // 1. NQG's nodes is subset of EQG's nodes
+        // -- already established via signature check
+        assert!(
+            existing_qg
+                .signature()
+                .is_weak_generalization_of(&new_qg.signature())
+        );
+
+        // Check if the queries are join compatible -- if the new query
+        // performs a superset of the joins in the existing query.
+        for (srcdst, ex_qge) in &existing_qg.edges {
+            match *ex_qge {
+                QueryGraphEdge::Join(_) => {
+                    if !new_qg.edges.contains_key(srcdst) { return None; }
+                    let new_qge = &new_qg.edges[srcdst];
+                    match *new_qge {
+                        QueryGraphEdge::Join(_) => {}
+                        // If there is no matching Join edge, we cannot reuse
+                        _ => return None,
+                    }
+                }
+                QueryGraphEdge::LeftJoin(_) => {
+                    if !new_qg.edges.contains_key(srcdst) { return None; }
+                    let new_qge = &new_qg.edges[srcdst];
+                    match *new_qge {
+                        QueryGraphEdge::LeftJoin(_) => {}
+                        // If there is no matching LeftJoin edge, we cannot reuse
+                        _ => return None,
```

Comment:
On that thought (though not relevant as a direct change here, and certainly applicable to all algorithms): we could totally reuse a `LeftJoin` as a plain `Join` by simply adding a filter that discards rows with NULLs in the right side columns. This is neat because filters add no materializations and can be queried through.